src/core/qgsdataitem.h

```diff
@@ -141,7 +141,22 @@ class CORE_EXPORT QgsDataItem : public QObject
     /** Returns the list of actions available for this item. This is usually used for the popup menu on right-clicking
      * the item. Subclasses should override this to provide actions.
      */
-    virtual QList<QAction *> actions() { return QList<QAction *>(); }
+    virtual QList<QAction *> actions() { return mActions; }
```

Comment:
I think it can be a bit misleading to have this method virtual with a default implementation based on a setter.
Why not just follow the current approach of overriding this in subclasses?

Reply:
Yes, looks to me like this could all be done by just overriding the existing actions() method.

If the long-term plans are to allow "injecting" new actions into existing items, then I think the correct approach would be to:

1. Rename existing QgsDataItem::actions() to QgsDataItem::createActions(), and make it protected (also add API break note to docs/api_break.dox)
2. Add a new (non virtual) method QgsDataItem::actions(). This would first get the list returned by createActions(), then add the extra actions present in mActions to the end of this list before returning it. A separator should be added between the actions returned by createActions() and those from mActions.
3. Rename setActions to "addActions", and in the api dox add a note along the lines of "Ownership of the actions remains unchanged, and callers must ensure the actions are appropriately parented to avoid leaking memory."

(ditto for menus)

Reply:
@nyalldawson @m-kuhn yes agree, at the moment I haven't found a condition where the action or menus need to be injected, thanks!